include/Components/InputFileDialog.hpp

```diff
@@ -0,0 +1,44 @@
+/*
+** Project FileShare, 2022
+**
+** Author Léo Lhuile
+**
+** Started on  Sat Jan 15 21:08:44 2022 Léo Lhuile
+** Last update Wed Sep 14 22:26:27 2022 Léo Lhuile
+**
+** InputFileDialog.hpp : Class to manage an InputFileDialog.
+*/
+
+#pragma once
+
+#include <TGUI/TGUI.hpp>
+
+namespace FileShare::GUI::Components {
+    class InputFileDialog : public tgui::EditBox {
+        public:
+            enum FileMode {
+                File,
+                Directory
+            };
+
+            InputFileDialog(const char* typeName = "Components::InputFileDialog", bool initRenderer = true);
+            ~InputFileDialog();
+
+            typedef std::shared_ptr<InputFileDialog> Ptr;
+			typedef std::shared_ptr<const InputFileDialog> ConstPtr;
+
+			static InputFileDialog::Ptr create() { return std::make_shared<InputFileDialog>(); }
+			static InputFileDialog::Ptr copy(InputFileDialog::ConstPtr widget) { return widget ? std::static_pointer_cast<InputFileDialog>(widget->clone()) : nullptr; }
+
+            void setMode(FileMode mode) { this->mode = mode; }
+            FileMode getMode() const { return this->mode; }
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<InputFileDialog>(*this); }
```

Comment:
Theses `typedef` and functions (ignoring the set/get mode) are repeated across all components, is it possible to use CRTP to define them in a template class that we can just inherit from and have them without code duplication ?

Reply:
These typedefs must be present in each component

https://tgui.eu/tutorials/latest-stable/custom-widgets/#basic-functionality

Reply:
I understand they need to be in each class, but the point of CRTP would be to define them in a template class that each class would inherit from, which would include the typedefs + functions in each class, and since it's template (with CRTP) it can resolve the proper types to the child class' type